api/lib/opentelemetry/context.rb

```diff
@@ -29,23 +33,50 @@ def current
         Thread.current[KEY] ||= ROOT
       end
 
-      # Sets the current context
-      #
-      # @param [Context] ctx The context to be made active
+      # @api private
       def current=(ctx)
         Thread.current[KEY] = ctx
       end
 
+      # Associates a Context with the caller's current Fiber. Every call to
+      # this operation should be paired with a corresponding call to detach.
+      #
+      # Returns a token to be used with the matching call to detach
+      #
+      # @param [Context] context The new context
+      # @return [Object] A token to be used when detaching
+      def attach(context)
+        prev = current
+        self.current = context
+        stack.push(prev)
+        stack.size
```

Comment:
There's an efficiency concern here because this requires 4 thread local lookups (reading `current`, writing `current`, and 2 `stack` accesses). It would be more efficient to only store the `stack` as a thread-local and have `current` implicitly stored at the top of the stack. (`detach` has a similar issue below)
```suggestion
        s = stack
        s.push(context)
        s.size
```